### PR TITLE
add day 9

### PR DESCRIPTION
**Pull request by djlf652**

### CODE REVIEW

File: src/09/sum_of_risk_levels.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        row.push(9)
+    })
+    
+    let output: number = 0
+
+    // Find the low points and add their risk levels to the total
+    for (let i = 1; i < input.length - 1; i++) {
+        for (let j = 1; j < input[i].length - 1; j++) {
+            if (
+                input[i][j] < input[i - 1][j] &&
```
Comment (Bekreth) [inline src/09/sum_of_risk_levels.ts:17]: Given that you use `input[i][j]` five times, I'd recommend assigning it to a variable. Possibly like `pointOfInterest`

File: src/09/sum_of_risk_levels.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
+export function sumOfRiskLevels(input: number[][]): number {
+    // Add padding to input
+    let yPadding: number[] = input[0].map(() => 9)
```
Comment (Bekreth) [inline src/09/sum_of_risk_levels.ts:3]: You have a small hiccup with your `yPadding` here.  You've pushed and unshifted the same object, so it ends up getting x-padded twice.  A simple fix would be to have the x-padding happen first.

Comment (Bekreth) [inline src/09/sum_of_risk_levels.ts:3]: This chuck is isolated from the following processing chunk and could be worth extracting into its own function.

File: test/09/largest_basins.test.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    test("when the map has 1 basin", () => {
+        const input: number[][] = [[1,2,9],
+                                   [2,9,9],
+                                   [9,9,9]]
+        expect(largestBasins(input)).toStrictEqual(3)
+    })
+    test("when the map has 3 basins", () => {
+        const input: number[][] = [[1,2,9,9],
+                                   [9,9,9,9],
+                                   [3,9,9,3],
```
Comment (Bekreth) [inline test/09/largest_basins.test.ts:13]: I'd suggest making this basic a bit larger to make sure you have a small test that show multiply happening.  (The current basin value is achieved if you'd added instead.

File: src/09/largest_basins.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
+function explore(input: number[][], point: number[], traveledPoints: number[][]): number[][] {
```
Comment (Bekreth) [inline src/09/largest_basins.ts:1]: I'd recommend putting this below the `largestBasin` function.  While C and C++ require declaration of functions to exist prior to their use, every modern language is quite a bit more flexible.  Putting this afterwards makes it read a bit more naturally: top to bottom.

Comment (Bekreth) [inline src/09/largest_basins.ts:1]: I'd recommend converting `point: number[]` to an interface.  Something like this
```typescript
interface Point {
  x: number,
  y: number
}
```

File: src/09/largest_basins.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    let adjacentPoints: number[][] = [
+        [i - 1, j],
+        [i + 1, j],
+        [i, j - 1],
+        [i, j + 1]
+    ]
+    adjacentPoints.forEach(point => {
+        let i: number = point[0]
+        let j: number = point[1]
+        if (input[i][j] < 9 && !traveledPoints.some(p => p[0] == i && p[1] == j)) {
```
Comment (Bekreth) [inline src/09/largest_basins.ts:14]: I'd recommend flipping this expression.  While in this problem your validity check is pretty simple (is the number 9) if that check is more complicated, you can spend a lot of time and effort solving for a solution you've already checked.

Comment (Bekreth) [reply src/09/largest_basins.ts:14]: `includes` is a better function for this problem than `some`

Comment (Bekreth) [inline src/09/largest_basins.ts:14]: As a bit of a pop-quiz, is this implementation a depth-first-search, or a breadth-first-search?

**Review by Bekreth — CHANGES_REQUESTED** [review]
Great job!  I think you can find a way to reduce code duplication.  There's an adage in software that goes "Don't Repeat Yourself".  Frequently you'll see it as DRY.

File: src/09/largest_basins.ts (hunk 5 above)
Comment (djlf652) [reply src/09/largest_basins.ts]: > `includes` is a better function for this problem than `some`

`includes` doesn't seem to work when the parameter is an array or an interface.

Comment (djlf652) [reply src/09/largest_basins.ts]: Depth-first-search

File: src/09/largest_basins.ts [hunk 6]
```diff
@@ -1,50 +1,35 @@
...
-    ]
-    adjacentPoints.forEach(point => {
-        let i: number = point[0]
-        let j: number = point[1]
-        if (input[i][j] < 9 && !traveledPoints.some(p => p[0] == i && p[1] == j)) {
-            traveledPoints = explore(input, point, traveledPoints)
-        }
-    })
-    return traveledPoints
+interface Point {
```
Comment (Bekreth) [inline src/09/largest_basins.ts:1]: `i` and `j` are quite canonically used for loop indexing.  In this case, a Point would be better served with geometric variables like `x` and `y`.

File: src/09/largest_basins.ts [hunk 7]
```diff
@@ -53,3 +38,27 @@ export function largestBasins(input: number[][]): number {
         .slice(0, 3)
         .reduce((previous, current) => previous * current, 1)
 }
+
+function yPadding(input: number[][]): number[] {
```
Comment (Bekreth) [inline src/09/largest_basins.ts:42]: Whoops, I think my prior comment was a bit unclear.  The suggestion was to move the entire padding (both X and Y) into a separate function so your code would look more like this
```typescript
function largestBasin(input: number[][]) {
  let paddedInput = padInput(input)
  let lowPoints: Point[] = []
  for (...) {
    ...
  }
}
```

From there, you could then put `function padInput(input: number[][])` into a stand alone file like `padder.ts` and import into both `largest_basins.ts` and `sum_of_risk_levels.ts` to DRY out your code.

File: src/09/largest_basins.ts (hunk 5 above)
Comment (Bekreth) [reply src/09/largest_basins.ts]: :partying_face:

Comment (Bekreth) [reply src/09/largest_basins.ts]: You're right.  I forgot about the issues TS has around equality checking.

**Review by Bekreth — APPROVED** [review]

**Pull request merged**